### PR TITLE
feat(logger): configurable through CLI & fix dupe provision

### PR DESCRIPTION
**Pull request by ocnc2**
## Summary by CodeRabbit

- **New Features**
  - Introduced enhanced logging configuration options, including time format, log level, and style settings.
  - Added new benchmarking tests for various logging levels and styles.

- **Improvements**
  - Updated logger initialization to utilize a more flexible configuration setup.
  - Enhanced configuration structures to support new logging settings.

- **Refactor**
  - Modified internal logger functions to improve configurability and maintainability.
  - Refactored logging methods to ensure consistent usage of the new logging configurations.

- **Chores**
  - Added new dependencies to support updated logging functionalities.

### CODE REVIEW

File: mod/config/config.go [hunk 1]
```diff
@@ -53,15 +55,17 @@ type Config struct {
 	Engine engineclient.Config `mapstructure:"engine"`
 	// KZG is the configuration for the KZG blob verifier.
 	KZG kzg.Config `mapstructure:"kzg"`
+	// Logger is the configuration for the logger.
+	Logger logger.Config `mapstructure:"logger"`
```
Comment (itsdevbear) [inline mod/config/config.go:59]: tbh I think this is overkill, should be configarable via the node-builder, but having it in a config file is too much

Comment (ocnc2) [reply mod/config/config.go]: yeah true i agree, was originally thinking some basic options such as time format could be configurable through cli

**Comment by itsdevbear** [comment]
oh i actually agree with the time thing, and color vs, no color vs json

File: mod/log/pkg/phuslu/config.go [hunk 2]
```diff
@@ -22,16 +22,67 @@ package phuslu
...
+
 // Config is a structure that defines the configuration for the logger.
 type Config struct {
 	// TimeFormat is a string that defines the format of the time in
 	// the logger.
-	TimeFormat string
+	TimeFormat string `mapstructure:"trace_color"`
+
+	// Colours for the different log levels.
+	TraceColor   string `mapstructure:"trace_color"`
```
Comment (itsdevbear) [inline mod/log/pkg/phuslu/config.go:47]: yeah remove colours and label, too overkill

File: mod/cli/pkg/components/logger.go [hunk 3]
```diff
@@ -33,15 +33,17 @@ import (
...
 	AppOpts servertypes.AppOptions
+	Out     io.Writer
 }
 
 // ProvideLogger creates a the default phuslu logger.
 // It reads the log level and format from the server context.
 func ProvideLogger(
 	in LoggerInput,
-) (log.Logger, error) {
+) log.Logger {
```
Comment (itsdevbear) [inline mod/cli/pkg/components/logger.go:44]: try to use our log.Logger[any] wherever you can

Comment (ocnc2) [reply mod/cli/pkg/components/logger.go:44]: not really sure why, but depinject is unable to resolve whenever i try to provide our `AdvancedLogger[any, sdk.Logger]` to the cosmos components...

despite our AdvancedLogger adhering to the same interface after applying generics.

File: mod/log/pkg/phuslu/config.go (hunk 2 above)
Comment (ocnc2) [reply mod/log/pkg/phuslu/config.go:47]: 👍

File: mod/log/pkg/phuslu/phuslu.go [hunk 4]
```diff
@@ -115,3 +116,41 @@
...
+func (l *Logger[ImplT]) WithConfig(cfg Config) *Logger[ImplT] {
+	l.logger.TimeFormat = cfg.TimeFormat
+	if cfg.Style == StylePretty {
+		l.useConsoleWriter()
+	} else if cfg.Style == StyleJSON {
+		l.useJSONWriter()
+	}
+	// todo: set verbosity
+	return l
+}
```
[the remark this answers is not in the document]
Comment (ocnc2) [reply mod/log/pkg/phuslu/phuslu.go]: 🥕 just take this

File: mod/node-core/pkg/builder/builder.go [hunk 5]
```diff
@@ -125,3 +130,7 @@ func (nb *NodeBuilder[NodeT]) Build(
 	}
 	return nb.node
 }
+
+func SetLoggerConfig(config *config.Config, logger log.Logger) {
+	logger.(*phuslu.Logger[log.Logger]).WithConfig(*config.GetLogger())
+}
```
[the remark this answers is not in the document]
Comment (ocnc2) [reply mod/node-core/pkg/builder/builder.go]: thanks

File: mod/config/pkg/template/template.go [hunk 6]
```diff
@@ -45,6 +45,16 @@ rpc-jwt-refresh-interval = "{{ .BeaconKit.Engine.RPCJWTRefreshInterval }}"
 # Path to the execution client JWT-secret
 jwt-secret-path = "{{.BeaconKit.Engine.JWTSecretPath}}"
 
+[beacon-kit.logger]
+# TimeFormat is a string that defines the format of the time in the logger.
+time-format = "{{.BeaconKit.Logger.TimeFormat}}"
+
+# Verbose is the level of verbose for the logger.
+verbose = "{{.BeaconKit.Logger.Verbose}}"
```
Comment (itsdevbear) [inline mod/config/pkg/template/template.go:53]: @ocnc2 can we rename this to verbosity?

**Review by itsdevbear — CHANGES_REQUESTED** [review]
nit

File: mod/log/pkg/phuslu/config.go [hunk 7]
```diff
@@ -20,18 +20,22 @@
...
-import "time"
-
 // Config is a structure that defines the configuration for the logger.
 type Config struct {
 	// TimeFormat is a string that defines the format of the time in
 	// the logger.
-	TimeFormat string
+	TimeFormat string `mapstructure:"time-format"`
+	// Verbose is the level of verbose for the logger.
+	Verbose bool `mapstructure:"verbose"`
```
Comment (itsdevbear) [inline mod/log/pkg/phuslu/config.go:29]: @ocnc2 convert this to the log level

"info", "trace", "debug" etc.

Comment (ocnc2) [reply mod/log/pkg/phuslu/config.go:29]: yep sounds good

**Pull request merged**